paddle/fluid/operators/sync_batch_norm_op.cu

```diff
@@ -0,0 +1,452 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include <algorithm>
+#include <cfloat>
+#include <string>
+#include <vector>
+#include "cub/cub.cuh"
+#include "paddle/fluid/framework/data_layout.h"
+#include "paddle/fluid/operators/batch_norm_op.h"
+#include "paddle/fluid/platform/cudnn_helper.h"
+#include "paddle/fluid/platform/float16.h"
+#include "paddle/fluid/platform/nccl_helper.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+using DataLayout = framework::DataLayout;
+template <typename T>
+using CudnnDataType = platform::CudnnDataType<T>;
+
+template <typename T, int BlockDim, framework::DataLayout layout>
+__global__ void KeLocalStats(const T *x, int N, int M, int C, T *mean_var) {
+  typedef cub::BlockReduce<T, BlockDim> BlockReduce;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+  T x_sum = 0;
+  T x2_sum = 0;
+  if (layout == framework::DataLayout::kNCHW) {
+    for (int i = threadIdx.x; i < N * M; i += blockDim.x) {
+      int n = i / M;
+      int id = n * C * M + blockIdx.x * M + i % M;
+      T x_in = x[id];
+      x_sum += x_in;
+      x2_sum += x_in * x_in;
+    }
+  } else {
+    for (int i = threadIdx.x; i < N * M; i += BlockDim) {
+      int id = i * C + blockIdx.x;
+      T x_in = x[id];
+      x_sum += x_in;
+      x2_sum += x_in * x_in;
+    }
+  }
+  __syncthreads();
+  T out = BlockReduce(temp_storage).Reduce(x_sum, cub::Sum());
+  __syncthreads();
+  if (threadIdx.x == 0) {
+    mean_var[blockIdx.x] = out / (N * M);
+  }
+  out = BlockReduce(temp_storage).Reduce(x2_sum, cub::Sum());
+  __syncthreads();
+  if (threadIdx.x == 0) {
+    mean_var[blockIdx.x + C] = out / (N * M);
+  }
+  if (blockIdx.x == 0 && threadIdx.x == 0) {
+    mean_var[2 * C] = static_cast<T>(1.0);
+  }
+}
+
+template <typename T>
+__global__ void KeSyncAndMovingStats(T *means, T *variances, T *num_dev,
+                                     const int C, const T momentum,
+                                     const double epsilon, T *sv_mean_data,
+                                     T *sv_inv_var_data, T *moving_means,
+                                     T *moving_variances) {
+  // sync stats across multi-devices
+  int gid = blockIdx.x * blockDim.x + threadIdx.x;
+  int stride = blockDim.x * gridDim.x;
+  for (int i = gid; i < C; i += stride) {
+    T mean = means[i] / (*num_dev);
+    T var = variances[i] / (*num_dev);
+    var = var - mean * mean;
```

Comment:
```
var = (sum((x - mean) ** 2)) / batch_size
var = (sum(x**2) - sum(2*x*mean) + sum(mean)**2) / batch_size
```
???

Reply:
@typhoonzero Here, use v = E(x^2) - E(x)^2 to calculate variance

Reply:
I see, thanks.